### PR TITLE
(PUP-8041) puppet device --resource

### PR DESCRIPTION
**Pull request by shermdog**
Added the ability to query resources via `puppet device`

### CODE REVIEW

**Comment by shermdog** [comment]
Will fixup commit message and all that once we get #6559 merged in, as we'll have to refactor for that.  Just putting in the PR now for review of overall flow, style, etc.

File: lib/puppet/application/device.rb [hunk 1]
```diff
@@ -212,27 +232,45 @@ def main
...
-          configurer.run(:network_device => true, :pluginsync => Puppet::Configurer.should_pluginsync?)
+          if options[:resource]
+            type = options[:resource][0]
+            Puppet.info _("retrieving resource: %{resource} from %{target} at %{scheme}%{url_host}%{port}%{url_path}") % { resource: type, target: device.name, scheme: scheme, url_host: device_url.host, port: port, url_path: device_url.path }
+
+            resources = find_resources(type, options[:resource][1])
+
+            if options[:to_yaml]
+              text = resources.map do |resource|
+                resource.prune_parameters(:parameters_to_include => @extra_params).to_hierayaml.force_encoding(Encoding.default_external)
```
Comment (thallgren) [inline lib/puppet/application/device.rb]: The method `#to_hierayaml` seems limited in how it produces the actual YAML. Is this to make something human readable only or is there some other purpose with the YAML output. Will the resource ever contain rich-data?

File: lib/puppet/application/device.rb [hunk 2]
```diff
@@ -212,27 +232,45 @@ def main
...
-          # ask for a ssl cert if needed, but at least
-          # setup the ssl system for this device.
-          setup_host
-
-          require 'puppet/configurer'
-          configurer = Puppet::Configurer.new
-          configurer.run(:network_device => true, :pluginsync => Puppet::Configurer.should_pluginsync?)
+          if options[:resource]
+            type = options[:resource][0]
+            Puppet.info _("retrieving resource: %{resource} from %{target} at %{scheme}%{url_host}%{port}%{url_path}") % { resource: type, target: device.name, scheme: scheme, url_host: device_url.host, port: port, url_path: device_url.path }
```
Comment (thallgren) [inline lib/puppet/application/device.rb]: nit pick, but why using double quotes on strings that doesn't contain any interpolation?

File: lib/puppet/application/device.rb (hunk 1 above)
Comment (shermdog) [reply lib/puppet/application/device.rb]: I copied that directly from puppet resource :)  I believe the feature is used to get data for Hiera.  We should probably consider adding in rich_data support to both `puppet resource` and `puppet device 
 --resource` in the future.

File: lib/puppet/application/device.rb (hunk 2 above)
Comment (shermdog) [reply lib/puppet/application/device.rb]: More copy-pasta from the existing bits here.  I can change it if that's preferred over what was already there.

**Comment by joshcooper** [comment]
@shermdog is checking which branch to target, don't merge until then.

**Comment by joshcooper** [comment]
ping @shermdog any updates on this?

**Comment by shermdog** [comment]
closed and re-opened to test appveyor

**Comment by joshcooper** [comment]
not sure what's up with appveyor, rekicked.

**Pull request merged**